### PR TITLE
Add Conditional Access Policy Support

### PR DESCRIPTION
**Pull request by alexwilcox9**
Based on the work in https://github.com/hashicorp/terraform-provider-azuread/pull/430 refactored to work with https://github.com/hashicorp/terraform-provider-azuread/pull/461 and also adding session controls and a few other attributes.

Let me know if you what you think and I'll crack on with the documentation

Thanks!

Closes: #348

### CODE REVIEW

**Comment by kaovd** [comment]
Nice work! I think while #430 OP seemed to be a bit distant not sure if it required forking for a whole new PR just yet. Everything passes my side so might as well put in some docs.

```
=== RUN   TestAccConditionalAccessPolicy_basic
=== PAUSE TestAccConditionalAccessPolicy_basic
=== RUN   TestAccConditionalAccessPolicy_complete
=== PAUSE TestAccConditionalAccessPolicy_complete
=== RUN   TestAccConditionalAccessPolicy_update
=== PAUSE TestAccConditionalAccessPolicy_update
=== CONT  TestAccConditionalAccessPolicy_basic
=== CONT  TestAccConditionalAccessPolicy_update
=== CONT  TestAccConditionalAccessPolicy_complete
--- PASS: TestAccConditionalAccessPolicy_basic (50.33s)
--- PASS: TestAccConditionalAccessPolicy_complete (52.10s)
--- PASS: TestAccConditionalAccessPolicy_update (170.42s)
PASS
ok      github.com/hashicorp/terraform-provider-azuread/internal/services/conditionalaccesspolicies     170.431s
```

**Comment by kaovd** [comment]
Just went ahead and tested this pre merge as 2.1 is coming up and looks good 

```
 env TF_ACC=1 go test -v -timeout=2h ./internal/services/conditionalaccesspolicies -run="TestAccConditionalAccessPolicy"                                                  [±ca-policies ✓]
=== RUN   TestAccConditionalAccessPolicy_basic
=== PAUSE TestAccConditionalAccessPolicy_basic
=== RUN   TestAccConditionalAccessPolicy_complete
=== PAUSE TestAccConditionalAccessPolicy_complete
=== RUN   TestAccConditionalAccessPolicy_update
=== PAUSE TestAccConditionalAccessPolicy_update
=== CONT  TestAccConditionalAccessPolicy_basic
=== CONT  TestAccConditionalAccessPolicy_update
=== CONT  TestAccConditionalAccessPolicy_complete
--- PASS: TestAccConditionalAccessPolicy_complete (136.32s)
--- PASS: TestAccConditionalAccessPolicy_basic (138.81s)
--- PASS: TestAccConditionalAccessPolicy_update (154.50s)
PASS
ok      github.com/hashicorp/terraform-provider-azuread/internal/services/conditionalaccesspolicies     154.506s

```

**Comment by manicminer** [comment]
Awesome thanks @kaovd. I have some minor changes to push tomorrow prior to merge but on the whole this is looking good 👍

**Comment by manicminer** [comment]
Unfortunately will have to push this to next week due to some testing issues

**Comment by alexwilcox9** [comment]
No problem, anything I can work on so it's ready for next week?

**Comment by manicminer** [comment]
@alexwilcox9 Thanks! Anything on the 2.2.0 milestone would be great :D

**Comment by manicminer** [comment]
No worries! I only got part way through reviewing this PR but at quick glance it looks pretty good. I'll circle back with any suggestions after today's release (and hopefully once we have fixed our testing issue).

**Comment by kaovd** [comment]
@alexwilcox9 @liammoat can I just check whats the actual difference between the beta endpoint and v1 upstream in SDK? Wherever possible want to use v1 if there is near enough feature parity.  Prehaps v1 didnt exist when originally made?

**Comment by kaovd** [comment]
@alexwilcox9 could you give this a rebase for the conflicts

File: docs/resources/conditional_access_policy.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+* `enabled` - (Optional) enables conditional access application control.
+
+---
+
+`persistent_browser` block supports the following:
+
+* `enabled` - (Optional) enables persistent browser session.
+* `mode` - (Optional) Defines which persistent browser mode is enabled. Possible values are: `always`, `never`.
+
+`sign_in_frequency` block supports the following:
```
Comment (manicminer) [inline docs/resources/conditional_access_policy.md:158]:
```suggestion

* `application_enforced_restrictions_enabled` - (Optional) Whether or not application enforced restrictions are enabled. Defaults to `false`.

-> Only Office 365, Exchange Online and Sharepoint Online support application enforced restrictions.

* `cloud_app_security_policy` - (Optional) Enables cloud app security and specifies the cloud app security policy to use. Possible values are: `mcasConfigured`, `monitorOnly`, `blockDownloads` or `unknownFutureValue`.
* `persistent_browser_mode` - (Optional) Specifies the persistent browser mode to use. Possible values are: `always`, `never`. Required when `persistent_browser_enabled` is specified.

-> `persistent_browser_mode` only works correctly when `included_applications` is set to `All`.

* `sign_in_frequency` - (Optional) Number of days or hours to enforce sign-in frequency. Required when `sign_in_frequency_period` is specified.
* `sign_in_frequency_period` - (Optional) The time period to enforce sign-in frequency. Possible values are: `hours` or `days`. Required when `sign_in_frequency_period` is specified.
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @alexwilcox9, thanks for this, this is really great work. I've made some comments primarily on the schema, and on moving the resource into the newly added `conditionalaccess` package. If you can take a look at these, then I'll review again in full and run it through some testing. Thanks!

**Review by manicminer — APPROVED** [review]
Thanks again for your work on this @alexwilcox9! LGTM! 🏎️ 

Test results:

Screenshot 2021-09-10 at 01 17 04

**Comment by alexwilcox9** [comment]
Thank you for all your help @manicminer, this has been a journey and an amazing introduction to working open source!

**Pull request merged**